chain/rust/src/assets.rs

```diff
@@ -0,0 +1,595 @@
+use cbor_event::{de::Deserializer, se::Serializer};
+use cml_core::{
+    error::{DeserializeError, DeserializeFailure, Key},
+    serialization::{fit_sz, Deserialize, Serialize, StringEncoding, LenEncoding, CBORReadLen},
+    ordered_hash_map::OrderedHashMap,
+};
+use cml_crypto::RawBytesEncoding;
+use std::{io::{BufRead, Seek, Write}};
+use std::cmp::PartialOrd;
+use crate::PolicyId;
+
+use std::collections::BTreeMap;
+
+use super::{AssetName, ScriptHash};
+
+pub type Coin = u64;
+
+/// Bundle of assets within range of T, grouped by PolicyID then AssetName
+#[derive(Clone, Debug, Default, PartialEq, serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+pub struct AssetBundle<T>(OrderedHashMap<PolicyId, OrderedHashMap<AssetName, T>>);
+
+impl<T> From<OrderedHashMap<PolicyId, OrderedHashMap<AssetName, T>>> for AssetBundle<T> {
+    fn from(bundle: OrderedHashMap<PolicyId, OrderedHashMap<AssetName, T>>) -> Self {
+        Self(bundle)
+    }
+}
+
+impl<T> std::ops::Deref for AssetBundle<T> {
+    type Target = OrderedHashMap<PolicyId, OrderedHashMap<AssetName, T>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T> std::ops::DerefMut for AssetBundle<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+/// Subtraction over a numerical type where the type's minimum is returned if underflow would occur.
+pub trait ClampedSub {
+    fn clamped_sub(&self, rhs: &Self) -> Self;
+}
+
+impl<T: num::CheckedSub + num::Bounded + num::Zero + Ord> ClampedSub for T {
+    fn clamped_sub(&self, rhs: &Self) -> Self {
+        self.checked_sub(rhs).unwrap_or_else(|| if *rhs < T::zero() { T::max_value() } else { T::min_value() })
+    }
+}
+
+impl<T> ClampedSub for AssetBundle<T>
+where T: num::CheckedAdd + num::CheckedSub + num::Zero + num::Bounded + Copy + Clone + ClampedSub {
```

Comment:
How do we feel about traits vs methods for consistency between rust/wasm vs idiomatic rust? Not just here, but in general.

Reply:
I feel like a lot of the point of this project was to allow idiomatic rust in the Rust file, so I would prefer idiomatic Rust instead of fine-tuning things for the WASM wrapper